tokio/src/sync/once_cell.rs

```diff
@@ -0,0 +1,310 @@
+                            "couldn't acquire a permit even though OnceCell value is uninitialized"
+                        );
+                    }
+                }
+            }
+        }
+
+        Err(AlreadyInitializedError)
+    }
+
+    /// Tries to initialize the value of the OnceCell using the async function `f`.
+    /// If the value of the OnceCell had already been initialized prior to this call
+    /// a reference to that initialized value is returned
+    ///
+    /// This will deadlock if `f` tries to initialize the cell itself.
+    pub async fn get_or_init<F, Fut>(&self, f: F) -> &T
+    where
+        F: FnOnce() -> Fut,
+        Fut: Future<Output = T>,
+    {
+        if self.initialized() {
+            // SAFETY: once the value is initialized, no mutable references are given out, so
+            // we can give out arbitrarily many immutable references
+            return unsafe { self.get_unchecked() };
+        } else {
+            // After acquire().await we have either acquired a permit while self.value
+            // is still uninitialized, or current thread is awoken after another thread
+            // has intialized the value and closed the semaphore, in which case self.initialized
+            // is true and we don't set the value here
+            match self.semaphore.acquire().await {
+                Ok(_permit) => {
+                    if !self.initialized() {
+                        let value = f().await;
+
+                        // SAFETY: There is only one permit on the semaphore, hence only one
+                        // mutable reference is created
+                        unsafe { self.set_value(value) };
+
+                        // SAFETY: once the value is initialized, no mutable references are given out, so
+                        // we can give out arbitrarily many immutable references
+                        return unsafe { self.get_unchecked() };
+                    } else {
+                        unreachable!("acquired semaphore after value was already initialized");
+                    }
+                }
+                Err(_) => {
+                    if self.initialized() {
+                        // SAFETY: once the value is initialized, no mutable references are given out, so
+                        // we can give out arbitrarily many immutable references
+                        return unsafe { self.get_unchecked() };
+                    } else {
+                        unreachable!(
+                            "semaphore acquire call failed when value not already intialized"
+                        );
+                    }
+                }
+            }
+        }
+    }
+}
+
+// OnceCell is Send (and Sync) because access to its UnsafeCell
+// is guarded by the semaphore permit and atomic operations
+// on `value_set`, as long as T itself is Send (and Sync)
+unsafe impl<T: Sync + Send> Sync for OnceCell<T> {}
+unsafe impl<T: Send> Send for OnceCell<T> {}
```

Comment:
This explanation seems incomplete. It doesn't explain why `Send` only requires `T: Send`, but `Sync` requires both `T: Send` and `T: Sync`.

Reply:
I don't know the exact reason for why we only need the `Sync` bound on `T` for `OnceCell<T>` to be Sync, I was assuming that inner generic types of a type for which we want to  implement `Sync` always require both bounds to hold. Can you explain, please?

Reply:
Well the definition of `Sync` is that a type is `Sync` if and only if it is safe to have immutable references to the value in multiple threads. In our case, we need both restrictions because:

 1. You can use an `&OnceCell<T>` to get an `&T`, so we need `T` to be `Sync`.
 2. Due to the `set` method operating on `&OnceCell<T>`, we can use it to _send_ values across threads too.

Strictly speaking the second requirement would only apply if you had a way to access the value in the `OnceCell` mutably, but it would make sense for us to have an `get_mut(&mut self) -> &mut T` and `into_inner(self) -> T` method anyway, so we do need this requirement.